conans/client/toolchain/cmake/base.py

```diff
@@ -33,8 +34,94 @@ class CMakeToolchainBase(object):
     filename = "conan_toolchain.cmake"
     project_include_filename = "conan_project_include.cmake"
 
-    _template_project_include = None
-    _template_toolchain = None
+    _toolchain_macros_tpl = textwrap.dedent("""
+        {% macro iterate_configs(var_config, action) -%}
+            {% for it, values in var_config.items() -%}
+                {%- set genexpr = namespace(str='') %}
+                {%- for conf, value in values -%}
+                    {%- set genexpr.str = genexpr.str +
+                                          '$<IF:$<CONFIG:' + conf + '>,"' + value|string + '",' %}
+                    {%- if loop.last %}{% set genexpr.str = genexpr.str + '""' -%}{%- endif -%}
+                {%- endfor -%}
+                {% for i in range(values|count) %}{%- set genexpr.str = genexpr.str + '>' %}{%- endfor -%}
+                {% if action=='set' %}
+                set({{ it }} {{ genexpr.str }})
+                {% elif action=='add_definitions' %}
+                add_definitions(-D{{ it }}={{ genexpr.str }})
+                {% endif %}
+            {%- endfor %}
+        {% endmacro %}
+        """)
+
+    _base_toolchain_tpl = textwrap.dedent("""
+        {% import 'toolchain_macros' as toolchain_macros %}
+
+        # Conan automatically generated toolchain file
+        # DO NOT EDIT MANUALLY, it will be overwritten
+
+        # Avoid including toolchain file several times (bad if appending to variables like
+        #   CMAKE_CXX_FLAGS. See https://github.com/android/ndk/issues/323
+        if(CONAN_TOOLCHAIN_INCLUDED)
+          return()
+        endif()
+        set(CONAN_TOOLCHAIN_INCLUDED TRUE)
+
+        {% block before_try_compile %}
```

Comment:
Maybe we could come up with a name that is more meaningful that ``before_try_compile``?